### ISSUE TITLE
Include web2 in main Twisted release

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11027]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| andrea reported|
|-|-|
|Trac ID|trac#1476|
|Type|enhancement|
|Created|2006-02-09 07:14:32Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1476 1476
type__enhancement enhancement
reporter__andrea andrea
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__web2 web2
keywords__ 
time__1139469272000000 1139469272000000
changetime__1150854813000000 1150854813000000
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__jknight cc__andrea cc__dreid
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
a recent change in setup.py (toplevel) broke web2 installation  
  
 http://www.cpushare.com/hg/Twisted?cs=f648811fdc61 
 
This will also install web2 by default (there's no risk in doing that AFIK) 
and I think it's beneficial if web2 gets more exposure.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
This is not a "bug"; it was a decision to not include web2 in the official
Twisted release stuff until everyone decided it's good to go.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
Try to install web2 on SVN without my change.  
   
I don't care how you fix it, but you have to fix it somehow. I fixed it in the   
way that was easiest for me to handle. 
 
Thanks.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Uh, yeah. "setup.py web2 install" used to work, and now is broken. This really has 
nothing to do with whether or not web2 is installed by default.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Your wish is granted.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
ok thanks SVN now can install web2 again. 
   
However it made it even more difficult for me to install twisted since pb, 
flow would also require special handling and even the supposed fix you checked 
into SVN broke my build scripts even more. I prefer to have a way to just run 
one command and install everything, 'all' is misleading word...   
   
So I added 'world', some project uses world to mean everything possible:   
 
http://www.cpushare.com/hg/Twisted/?cs=52049b455442
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
flow is deprecated.  pb is experimental.  just like web2, these should not be
installed by default currently.  if you want to use them, you have to say you
want to use them.  "all" isn't misleading, because it is really installing all
parts of twisted which are intended to be used.  note that it also does nothing
with the "win32" or "admin" directories: like pb and web2, these exist for the
Twisted developers, not end users.

Sorry if this is inconvenient for you, but it saves us time by avoiding
questions from people less experienced with twisted.

I don't really want to make setup.py even more complex than it is already. 
Clearly it is already so complex that seemingly innocuous changes actually break
behavior some people are relying on.  To me, this says "unit tests needed".
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
Where is it written that flow is deprecated? This is the first time I hear           
about it...        
           
http://twistedmatrix.com/documents/current/api/twisted.html          
          
no sign of deprecation in the docs, and neither in the code.         
         
cd ~/devel/python/network/Twisted/twisted/flow/         
grep -n -r Deprecation . /dev/null         
         
weirdness (but I'm not surprised anymore)  
         
Anyway I agree about removing pb from world is reasonable but I couldn't       
imagine flow was overnight-deprecated too, it's pure luck I didn't use it yet!   
 
So feel free to leave SVN as is, but I'll just add 'web2' to 'all' in my tree  
since I totally depend on web2. (I know you don't consider web2 good enough,        
but it's orders of magnitude better than web + nevow already, it allows a       
beautiful integration with Cheetah and FormEncode (it'll be even better as 
soon as Cheetah and FormEncode will speak deferreds, I checked the code and I 
think it's possible to do that with reentrance and nesting, with finegrined 
control as well on the parallelism of the async callbacks, then nevow is 
dead), so it's a much safer codebase to work on than the nevow CPU hog (with 
CPU hog I mean the fact I'm measured that Cheetah/FormEncode combination 
renders the same html between 5 to 10 times than Nevow depending on the page 
[live is only a x4 boost because most time is spent in the database sql 
queries and not so much in the html rendering, most of the boost of klive is 
thanks to the sql scaling in SMP], so it's an immense worthwhile improvement 
if one has an hope to get any traffic ever, even if you ignore the fact 
Cheetah it's so much easier to code with).
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
I'm going offtopic... but the livepage stuff should be extracted from nevow   
and become a library to plug on top of web2.   
  
As for ajax, mochikit should make it on top of web2 too (I used formencode,  
since the last thing I want to deal with is javascript, no matter in which  
share or form ;). If one day I'll have to make the page look nicer, mochikit  
is the one I will use.  
  
I'm not going to hide behind twistedmatrix.com and divmod.org when there are  
projects like subway and turbogears doing the right thing, I want twisted (at  
least my version) to join the python web community at large, that already  
solved those problems better than nevow ever did.  
  
All code is shareable (it'll interface more cleanly once it'll learn what a  
deferred is), and thanks to my usage of twisted.web2 instead of cherrypy I can  
use spread and other twisted features that I couldn't use if I were to use 
cherrypy or webware. Doing inter process communication using synchronous sql 
isn't the most scalable thing in the world, so I'm optimistic this design with 
twisted.web2 is more optimal that using an unmodified subway.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
http://www.larsen-b.com/Article/216.html  
  
The article is totally right here: " Anyways, the choice is really hard  
because it's time to choose right now, and I believe that some frameworks will  
die sooner than others.."  
  
I totally agree with him and that's very bad.  
  
The reason projects like Subway and Turbogears are extremely positive, is that  
they tell the world which libs people should use to render html with python.  
That is FormEncode/Mochikit/Cheetah and some other. Even if Turbogears and  
Subway are an unnecessary duplication of efforts, as long as they almost 
completely overlap in the code like today that's ok.  
 
Thanks to projects like Subway and Turbogears, the risk of those core libs 
becoming obsolete overnight will decrease and they will help to make the other  
solutions obsolete quicker.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Flow and Pair are both listed here as "unmaintained":

http://twistedmatrix.com/projects/

There are no deprecation warnings in the code because deprecation implies that
there will be a replacement - there will be no replacement, those projects are
just generally not useful and there is nobody to maintain them.  Both of their
authors have moved on.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
unmaintained has nothing to do with a package being obsolete.   
   
There are tons of drivers in the kernel that are unmaintained (in the  
maintainers file the unmaintained but not obsolete packages are defined  
orphan).  
 
So for not native english speakakers like me, calling something unmaintained  
means you can use it, except you may have to maintain it yourself if you do.  
  
Going obsolete means something has obsoleted it and you should not use it.  
  
They're very different things, so please clarify if this is obsolete or 
unmaintained, thanks. 
 
Apologies for going partly offtopic after hearing that flow was 
overnight-obsolete incidentally despite no document ever documented it as 
obsolete.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
btw, the exact word used was deprecated not obsoleted, but that still the  
basic message you send is that flow must not be used. 
  
I see no word anywhere that says that you must not use it, unmaintained means  
you can use it if you maintain it yourself.  
  
Also note, there is a replacement for flow, that is change the code not to use 
flow. It's just a different API. So if you think flow is deprecated because it 
generates substantially worse code than non-flow standard deferred API, then 
adding deprecation warnings is a very good idea. But such a choice must be 
motivated by technical arguments like flow generating worse code, not by the 
simple fact currently there is no maintainer. If a design is good, it's still 
good even without a maintainer for it. It's not like code can get worse if you 
don't change it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
You're right about the abruptness of the disappearance of twisted.flow.  This
has been discussed on IRC and the mailing list, but never actually indicated in
the code itself.  I've checked in a change to the doc string and added a
PendingDeprecationWarning to the package.  I've also re-added flow to the 'all'
group of subprojects.  We'll leave it there for at least one release before
dropping it.

I said "deprecated" for lack of a better word.  There is no direct replacement
for twisted.flow.  The case is more that twisted.flow does not represent a way
of programming which many (if any) current Twisted developers feel is
advantageous.  In addition to this, it is not actively maintained (this is
evident by the fact that it has had several intermittently failing unit tests
for at least a year) I think "deprecated" comes closest to the meaning I was
trying to convey.  This is the general Twisted policy.  We do not remove things
that are unmaintained, only things that are clearly detrimental and have no hope
of improving.  I hope this clarifies things a bit.

Regarding the "Goodness" of web2, I don't think this is the appropriate scale to
judge things on.  There's definitely lots of bad stuff (along with the totally
awesome stuff) in the parts of Twisted that are marked "stable" and will remain
in the release for a long time to come.  The issue is primarily the likelihood
that web2 will change in a way which will break applications using it.  In the
past, Twisted releases have included highly unstable code and marked it as such.
 However, the project as a whole is now more mature and it can be expected to
provide greater stability in this area.  Requiring web2 to be installed
explicitly is a way of indicating this to users in a way which is pretty
difficult to miss.  If you have a suggestion for a better way, I'm all ears.

Regarding Nevow and LivePage, you should take that discussion up on the
twisted-web or divmod-dev mailing list.  You might not want to bother, though,
unless you are willing to contribute the time to implement the changes.  It's
entirely possible this will happen eventually without direct contributions from
you, but I for one and not going to bother until I am happy with the underlying
platform.  If this means you are going to stop using Twisted (I can't quite
figure out how to interpret your "hide behind ..." comments), that's your choice.

As for joining the community, it sounds like you think the best way to
accomplish that is to completely eradicate Nevow from the face of the planet. 
While I can understand that sentiment (and if you wish to pursue it I will not
interfere unduly), I don't see why the very existence of Nevow is an impediment
to anything.  You have demonstrated that anyone who is unhappy with it can
continue to use Twisted in combination with some other high-level web system. 
That seems like the best kind of community participation.  Best of all, if
libraries outside of the Twisted sphere gain some notion of Deferreds, it seems
like everyone comes out a winner.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
Ok, I think the crucial point is "does not represent a way         
of programming which many (if any) current Twisted developers feel is         
advantageous". So I really appreciate you added the deprecation warnings         
giving the above message to the user "i.e. not advantageous -> use the         
standard deferreds instead". I guess the fact it's unmaintained is just a side         
effect of the basic issue that it was already informally considered         
deprecated. So this point is solved.      
         
About livepage, I'm not going to commit any work on it myself today, simply       
because I want my web to be compatible with all browsers (livepage for example       
doesn't work well under konqueror, and there are older browsers that may not         
support it like lynx ;). The advantage that new javascript related       
technologies like  livepage and AJAX give to my webpages aren't enough to take       
any risk of incompatibilities. However I welcome AJAX (Mochikit) and livepage,  
I've no doubt in the future I could be using it too.       
       
I totally agree about everything else you said, twisted.web2 must allow      
everything to work on top of it, be it nevow (please keep the compatibility      
stuff isolated like now), twcgi, cheetah, forms or formencode or a standalone      
livepage. I'd be happy to see nevow eventually taking over the web in the      
long term, it's certainly possible to fix nevow to perform better and be more      
flexible. My only point is that it will require some fundamental change and I   
don't see why the twisted community should spend time fixing nevow instead of   
using already existing products like the ones you can find in subway and   
turbogears.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @radix|
|-|-|

radix is the release guy, although I don't think we actually need this ticket.  When it is time to include web2 in a Twisted release, I'm pretty sure we'll know it.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set status to `closed`|
|-|-|

Agreed; closing.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
